src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -171,5 +165,142 @@ private static DateTime CreateDateTimeFromSystemTime(in Interop.Kernel32.SYSTEMT
 
             return (delegate* unmanaged[SuppressGCTransition]<ulong*, void>)pfnGetSystemTime;
         }
+
+        private static unsafe DateTime UpdateLeapSecondCacheAndReturnUtcNow()
+        {
+            // From conversations with the Windows team, the OS has the ability to update leap second
+            // data while applications are running. Leap second data is published on WU well ahead of
+            // the actual event. Additionally, the OS's list of leap seconds will only ever expand
+            // from the end. There won't be a situation where a leap second will ever be inserted into
+            // the middle of the list of all known leap seconds.
+            //
+            // Normally, this would mean that we could just ask "will a leap second occur in the next
+            // 24 hours?" and cache this value. However, it's possible that the current machine may have
+            // deferred updates so long that when a leap second is added to the end of the list, it
+            // actually occurs in the past (compared to UtcNow). To account for this possibility, we
+            // limit our cache's lifetime to just a few minutes (the "validity window"). If a deferred
+            // OS update occurs and a past leap second is added, this limits the window in which our
+            // cache will return incorrect values.
+            //
+            // We don't ever expect FileTimeToSystemTime or SystemTimeToFileTime to fail, but in theory
+            // they could do so if the OS publishes a leap second table update to all applications while
+            // this method is executing. If the time conversion routines fail, we'll re-run this method's
+            // logic from the beginning.
+
+            Debug.Assert(s_systemSupportsLeapSeconds);
+            Debug.Assert(LeapSecondCache.ValidityPeriodInTicks < TicksPerDay - TicksPerSecond, "Leap second cache validity window should be less than 23:59:59.");
+
+        TryAgain:
+
+            ulong fileTimeNow;
+            s_pfnGetSystemTimeAsFileTime(&fileTimeNow);
+
+            // If we reached this point, our leap second cache is stale, and we need to update it.
+            // First, convert the FILETIME to a SYSTEMTIME.
+
+            Interop.Kernel32.SYSTEMTIME systemTimeNow;
+            ulong hundredNanoSecondNow = fileTimeNow % TicksPerMillisecond;
+
+            // We need the FILETIME and the SYSTEMTIME to reflect each other's values.
+            // If FileTimeToSystemTime fails, call GetSystemTime and try again until it succeeds.
+            while (Interop.Kernel32.FileTimeToSystemTime(&fileTimeNow, &systemTimeNow) == Interop.BOOL.FALSE)
+            {
+                goto TryAgain;
```

Comment:
If this loops forever for some reason -- I guess we assume the OS is broken and it doesn't really matter that we didn't fail?

Reply:
I originally had `Environment.FailFast` in here but figured that there _might_ be a race condition where the OS is publishing new data at the same time we're running this method. If something's borked in the OS we'll probably loop forever. Hopefully we'll only loop until midnight (which is 2 seconds away at worst). :)